### PR TITLE
Stats 2: Avg. Resp Time, structured JSON with topics, types, totals

### PR DESCRIPTION
**Pull request by joshleecreates**
Example output:
```
{"totals":{"total":2,"nack_resp_time":5.3e4,"nack":1,"mean_resp_time":1.1705e6,"ignore_resp_time":null,"ignore":0,"ack_resp_time":2.288e6,"ack":1},"topics":{"events":{"test.unhappy":{"total":1,"nack_resp_time":5.3e4,"nack":1,"mean_resp_time":5.3e4,"ignore_resp_time":null,"ignore":0,"ack_resp_time":null,"ack":0},"apiv1.user.update":{"total":1,"nack_resp_time":null,"nack":0,"mean_resp_time":2.288e6,"ignore_resp_time":null,"ignore":0,"ack_resp_time":2.288e6,"ack":1},"totals":{"total":2,"nack_resp_time":5.3e4,"nack":1,"mean_resp_time":1.1705e6,"ignore_resp_time":null,"ignore":0,"ack_resp_time":2.288e6,"ack":1}}}}
```

### CODE REVIEW

File: lib/exque/status.ex [hunk 1]
```diff
@@ -21,37 +21,135 @@ defmodule Exque.Status do
...
+      start: System.os_time(),
+      topic: message.metadata.topic,
+      type: message.metadata.type,
+    })
+    {:noreply, Map.put(state, :messages, messages)}
+  end
+
+  defp filter_messages(messages, filter, term) do
+    messages
+    |> Enum.reject(fn({_, message}) -> (Map.get(message, filter) != term) end)
```
Comment (dmvt) [inline lib/exque/status.ex]: without the pipe since it's only 1 call

File: lib/exque/status.ex [hunk 2]
```diff
@@ -21,37 +21,135 @@ defmodule Exque.Status do
...
+  end
+
+  defp get_duration([]), do: nil
+  defp get_duration(messages) do
+    sum = messages
+      |> Enum.map(fn({_key, value}) -> value.end - value.start end)
+      |> Enum.reduce(fn(duration, acc) -> acc + duration end)
+    sum / Enum.count(messages)
```
Comment (dmvt) [inline lib/exque/status.ex]: you said "that's not right... that shouldn't be dividing"

File: lib/exque/status.ex [hunk 3]
```diff
@@ -21,37 +21,135 @@ defmodule Exque.Status do
...
-    )
-    {:noreply, new_state}
+      |> Enum.into(%{})
+  end
+
+  # new messages & cached stats
+  defp index(%{messages: messages, stats: stats} = state) do
+    new_stats = %{totals: totals(messages), topics: get_stats(messages)}
+    all_stats = Map.merge(stats, new_stats, &stats_merge/3)
+    new_state = state |>  Map.merge(%{stats: all_stats}) |> Map.drop([:messages])
```
Comment (dmvt) [inline lib/exque/status.ex]: `new_state = Map.merge(state, %{messages: %{}, stats: all_stats})`

File: lib/exque/status.ex [hunk 4]
```diff
@@ -21,37 +21,135 @@ defmodule Exque.Status do
...
+  end
+  defp inject_mean_response_times(map) when is_map(map) do
+    map
+    |> Enum.map(fn({k, v}) -> {k, inject_mean_response_times(v)} end)
+    |> Enum.into(%{})
+  end
+  defp inject_mean_response_times(value), do: value
+
+  defp stats_merge(_, first, second) when is_map(first),
+    do: Map.merge(first, second, &stats_merge/3)
```
Comment (dmvt) [inline lib/exque/status.ex]: need a nil check for `is_map(first)`

File: lib/exque/status.ex [hunk 5]
```diff
@@ -21,37 +21,135 @@ defmodule Exque.Status do
...
+  defp stats_merge(_, first, second) when is_integer(first), do: first+second
+  defp stats_merge(_, first, nil) when is_integer(first), do: first
+  defp stats_merge(_, _first, second), do: second
+
+  defp topic_stats(topic, messages) do
+    topic
+    |> Enum.map(&({&1, totals(filter_messages(messages, :type, &1))}))
+    |> Enum.into(%{})
```
Comment (dmvt) [inline lib/exque/status.ex]: Let's make this a reduce for the efficiencies

File: lib/exque/status.ex [hunk 6]
```diff
@@ -21,37 +21,144 @@ defmodule Exque.Status do
...
+      messages,
+      fn({_, message}) -> (Map.get(message, filter) == term) end
     )
-    {:noreply, new_state}
+  end
+
+  defp get_duration([]), do: nil
+  defp get_duration(messages) do
+    messages
+    |> Enum.map(fn({_key, value}) -> value.end - value.start end)
```
Comment (joshleecreates) [inline lib/exque/status.ex]: This *could* blow up if it gets hit while a message is still being processed. Needs to be a map.get with a default value for value.end

File: lib/exque/status.ex [hunk 7]
```diff
@@ -21,37 +21,144 @@ defmodule Exque.Status do
...
+    })
+  end
+  defp inject_mean_response_times(map) when is_map(map) do
+    map
+    |> Enum.map(fn({k, v}) -> {k, inject_mean_response_times(v)} end)
+    |> Enum.into(%{})
+  end
+  defp inject_mean_response_times(value), do: value
+
+  defp stats_merge(_, first, nil), do: first
```
Comment (joshleecreates) [inline lib/exque/status.ex]: This is actually not necessary because the merge function is only called to resolve conflicts

**Pull request merged**